AnnService/inc/Core/Common/FineGrainedLock.h

```diff
@@ -16,43 +16,27 @@ namespace SPTAG
         class FineGrainedLock {
         public:
             FineGrainedLock() {
-                rwlock.reset(new std::shared_timed_mutex);
-            }
-            ~FineGrainedLock() { 
-                for (size_t i = 0; i < locks.size(); i++)
-                    locks[i].reset();
-                locks.clear();
-            }
-            
-            void resize(SizeType n) {
-                SizeType current = (SizeType)locks.size();
-                if (current <= n) {
-                    {
-                        std::unique_lock<std::shared_timed_mutex> lock(*rwlock);
-                        locks.resize(n);
-                    }
-                    for (SizeType i = current; i < n; i++)
-                        locks[i].reset(new std::mutex);
-                }
-                else {
-                    for (SizeType i = n; i < current; i++)
-                        locks[i].reset();
-                    locks.resize(n);
-                }
+                m_locks.reset(new std::mutex[m_poolSize + 1]);
             }
+            ~FineGrainedLock() {}
 
             std::mutex& operator[](SizeType idx) {
-                std::shared_lock<std::shared_timed_mutex> lock(*rwlock);
-                return *locks[idx];
+                unsigned index = hash_func((unsigned)idx);
+                return m_locks[index];
             }
 
             const std::mutex& operator[](SizeType idx) const {
-                std::shared_lock<std::shared_timed_mutex> lock(*rwlock);
-                return *locks[idx];
+                unsigned index = hash_func((unsigned)idx);
+                return m_locks[index];
             }
         private:
-            std::unique_ptr<std::shared_timed_mutex> rwlock;
-            std::vector<std::shared_ptr<std::mutex>> locks;
+            static const int m_poolSize = 16383;
```

Comment:
It's abnormal to declare it as m_ as actually it's a static constant one which can not be modified, it's clearer as PoolSize